src/library_dyncall.js

```diff
@@ -0,0 +1,67 @@
+mergeInto(LibraryManager.library, {
+  {{{ (function() { global.wbind = function() { return SHRINK_LEVEL == 0 ? 'wbind' : 'wasmTable.get'; }; return null; })(); }}}
+  {{{ (function() { global.getDynCaller = function(sig) { return MINIMAL_RUNTIME ? `dynCalls[${sig}]` : `Module["dynCall_${sig}]`; }; return null; })(); }}}
+
+#if SHRINK_LEVEL == 0
+  // A mirror copy of contents of wasmTable in JS side, to avoid relatively
+  // slow wasmTable.get() call. Only used when not compiling with -Os or -Oz.
+  _wasmTableMirror: [],
+
+  $wbind__deps: ['_wasmTableMirror'],
+  $wbind: function(funcPtr) {
+    var func = __wasmTableMirror[funcPtr];
+    if (!func) {
+      if (funcPtr >= __wasmTableMirror.length) __wasmTableMirror.length = funcPtr + 1;
+      __wasmTableMirror[funcPtr] = func = wasmTable.get(funcPtr);
+    }
+    return func;
+  },
+
+  $dynCall__deps: ['$wbind'],
+  $bindDynCall__deps: ['$wbind'],
+  $wbindArray__deps: ['$wbind'],
+#else
+  $wbind: function(funcPtr) {
+    // In -Os and -Oz builds, do not implement a JS side wasm table mirror for small
+    // code size, but directly access wasmTable, which is a bit slower.
+    return wasmTable.get(funcPtr);
+  },
+#endif
+
+  // A helper that binds a wasm function into a form that can be called by passing all
+  // the parameters in an array, e.g. wbindArray(func)([param1, param2, ..., paramN]).
+  $wbindArray: function(funcPtr) {
+    var func = {{{wbind()}}}(funcPtr);
+    return func.length
+      ? function(args) { return func.apply(null, args); }
+      : function() { return func(); }
+  },
+
+  // A helper that returns a function that can be used to invoke function pointers, i.e.
+  // getDynCaller('vi')(funcPtr, myInt);
+  $getDynCaller: function(sig, funcPtr) {
+    return {{{getDynCaller('sig')}}};
+  },
+
+  $bindDynCall: function(sig, funcPtr) {
+    // For int64 signatures, use the dynCall_sig dispatch mechanism.
+    if (sig.includes('j')) return function(args) {
+      return {{{getDynCaller('sig')}}}.apply(null, [funcPtr].concat(args));
+    }
+    // For non-int64 signatures, invoke via the wasm table.
+    var func = {{{wbind()}}}(funcPtr);
+    return func.length
+      ? function(args) { return func.apply(null, args); }
+      : function() { return func(); }
+  },
+
+  $dynCall: function(sig, funcPtr, args) {
+    // For int64 signatures, use the dynCall_sig dispatch mechanism.
+    if (sig.includes('j')) {
+      return {{{getDynCaller('sig')}}}.apply(null, [funcPtr].concat(args));
```

Comment:
The convention in other library files seem to be to have white space inside the `{{{ ... }}}`

Reply:
Can use just implement this in terms if `bindDynCall` above?   

```
$dynCall: function(sig, funcPtr, args) { return bindDynCall(sig, funcPtr).apply(null, args); }
```

I imagine it would be a slight perf loss but a code size win,  and I don't think the generic runtime dynCall like this should be on anyones critical path, right?

Reply:
I don't think there is reason to make this code suboptimal, it is just a few lines. But added a variant in -Os/-Oz to reuse the code to make it smaller.